esp-hal/src/interrupt/mod.rs

```diff
@@ -135,19 +135,50 @@ pub trait InterruptConfigurable: crate::private::Sealed {
 pub struct InterruptHandler {
     f: extern "C" fn(),
     prio: Priority,
+    #[cfg(riscv)]
+    nested: bool,
 }
 
 impl InterruptHandler {
     /// Creates a new [InterruptHandler] which will call the given function at
     /// the given priority.
     pub const fn new(f: extern "C" fn(), prio: Priority) -> Self {
-        Self { f, prio }
+        Self {
+            f,
+            prio,
+            #[cfg(riscv)]
+            nested: true,
+        }
+    }
+
+    /// Creates a new [InterruptHandler] which will call the given function at
+    /// the given priority with disabled interrupt nesting.
+    ///
+    /// Usually higher priority interrupts get served while handling an interrupt.
+    /// Using this the interrupt handler won't get preempted by higher priority interrupts.
+    #[cfg(riscv)]
+    pub fn new_not_nested(f: extern "C" fn(), prio: Priority) -> Self {
+        Self {
+            f,
+            prio,
+            nested: false,
+        }
     }
 
-    /// The function to be called
+    /// The function to be called.
+    #[cfg_attr(
+        riscv,
+        doc = "\n\nNote that the function pointer might be misaligned. Don't ever just call the function."
+    )]
```

Comment:
I'll not block this PR on this, but we really, really should return a different type that prevents simply calling the interrupt handler. The user can call the function by its name anyway (or at least they are supposed to be). We don't need the value we return from here to be callable, or we can add logic to ensure it in case we do.

Reply:
Thanks for your understanding - I was considering changing it in the PR which adds a lot of noise, but doing it in a follow up PR is a good thing to do. Will take care of that